src/app/views/forgotten-password/setForgottenPasswordController.jsx

```diff
@@ -11,12 +15,115 @@ export class SetForgottenPasswordController extends Component {
+            password: this.state.password
+        };
+
+        this.setState({ isSubmitting: true, showInputError: false }, () => {
+            this.requestPasswordChange(requestBody);
+        });
+    }
+
+    requestPasswordChange(body) {
+        user.setForgottenPassword(body)
+            .then(() => {
+                this.setState({
+                    hasSubmitted: true,
+                    isSubmitting: false
+                });
+            })
+            .catch(error => {
+                this.handlePasswordResetError(error);
+                this.setState({
+                    isSubmitting: false
+                });
+            });
+    }
+
+    handlePasswordResetError(error) {
+        const notification = {
+            type: "warning",
+            isDismissable: true,
+            autoDismiss: 15000
+        };
+
+        if (error != null && error.status != null) {
+            if (error.status === 400) {
+                // All validation errors will be captured by this; if using the web interface validation is checked before you can submit
+                console.error("Unable to validate the type, email, password, or verification_token in the request");
+                log.event("Unable to validate the type, email, password, or verification_token in the request", log.error(error));
+                notification.message = errCodes.RESET_PASSWORD_VALIDATION_ERR;
+            } else if (error.status === 500) {
+                console.error("Invalid request body");
+                log.event("Invalid request body", log.error(error));
+                notification.message = errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR;
+            } else if (error.status === 501) {
+                console.error("Requested unimplemented password change type");
+                log.event("Requested unimplemented password change type", log.error(error));
+                notification.message = errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR;
+            } else {
+                console.error(errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR);
+                log.event(errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR, log.error(error));
+                notification.message = errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR;
+            }
+        } else {
+            console.error(errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR);
+            log.event(errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR, log.error(error));
+            notification.message = errCodes.RESET_PASSWORD_REQUEST_UNEXPECTED_ERR;
+        }
+        notifications.add(notification);
+    }
+
+    validityCheck(isValid, password) {
+        // Only show input error if user had previously tried to submit password and it is still invalid
+        const showInputError = !isValid && this.state.showInputError;
+        this.setState({
+            passwordIsValid: isValid,
+            password: password,
+            showInputError: showInputError
```

Comment:
I think this can be simplify too. you don't need two properties `passwordIsValid ` and separate `showInputError` they represent the same situation. If the password is not valid show the error. you don't show the error when password is valid. state must be as simple as possible and only contain elements which change need trigger re render.

Reply:
Ah, so these two are actually storing two different things,
`passwordIsValid` is storing whether or not the password is valid, this is only used to track if we can submit the password to the server.
`showInputError` doesn't track whether the password is correct or not. It instead tracks if the user has attempted to submit the password before meeting the validation rules, and then tracks whether the user has corrected these issues.
To put it another way, 
- `passwordIsValid` is false from the start, `showInputError` is also false from the start.
- if the user types `passwordIsValid` may turn true, whereas `showInputError` remains as false
- If the user submits but the password is invalid, then `showInputError` will then turn true (so only after the user clicks the `Confirm password` button) It will then remain true until they fix all validation issues.

If we used just `passwordIsValid` then as soon as the page loads up the password field would be red, and this can cause issues in accessibility for 'anxious' users as we are immediately telling them something is wrong.

Reply:
Hi Daniel, if you fancy I can help you with this. you really do not need those two states. errors can be activated as a result of derived state from validation. But it is not big deal, just worth keeping the state as simple as possible.

Reply:
Yes, I think it would be worth a conversation on this on Monday, if you are up for that :)

Reply:
Have moved the `passwordIsValid` to attribute in class rather than state var